packages/astro/src/virtual-modules/i18n.ts

```diff
@@ -191,3 +201,161 @@ export const getPathByLocale = (locale: string) => I18nInternals.getPathByLocale
+ *
+ * @param {APIContext} context The context passed to the middleware
+ * @param {ValidRedirectStatus?} statusCode An optional status code for the redirect.
+ */
+export const redirectToDefaultLocale =
+	i18n?.routing === 'manual'
+		? I18nInternals.redirectToDefaultLocale({
+				base,
+				trailingSlash,
+				format,
+				defaultLocale,
+				locales,
+				routing: strategy,
+				domains,
+				fallback,
+			})
+		: noop('redirectToDefaultLocale');
+
+/**
+ *
+ * Use this function to return a 404 when:
+ * - the current path isn't a root. e.g. / or /<base>
+ * - the URL doesn't contain a locale
+ *
+ * When a `Response` is passed, the new `Response` emitted by this function will contain the same headers of the original response.
+ *
+ * @param {APIContext} context The context passed to the middleware
+ * @param {Response?} response An optional `Response` in case you're handling a `Response` coming from the `next` function.
+ *
+ */
+export const notFound =
+	i18n?.routing === 'manual'
+		? I18nInternals.notFound({
+				base,
+				trailingSlash,
+				format,
+				defaultLocale,
+				locales,
+				routing: strategy,
+				domains,
+				fallback,
+			})
+		: noop('notFound');
+
+/**
+ * Checks whether the current URL contains a configured locale. Internally, this function will use `APIContext#url.pathname`
+ *
+ * @param {APIContext} context The context passed to the middleware
+ *
+ *
+ */
+export const requestHasLocale =
+	i18n?.routing === 'manual' ? I18nInternals.requestHasLocale(locales) : noop('requestHasLocale');
+
+export const useFallback: UseFallback =
+	i18n?.routing === 'manual'
+		? I18nInternals.useFallback({
+				base,
+				trailingSlash,
+				format,
+				defaultLocale,
+				locales,
+				routing: strategy,
+				domains,
+				fallback,
+			})
+		: noop('useFallback');
+
+// @ematipico: This is a helper type because, because I don't know how to extract from TS
+type Options = {
+	prefixDefaultLocale?: boolean;
+	redirectToDefaultLocale?: boolean;
+};
+/**
+ * @param {AstroConfig['i18n']} customOptions
+ *
+ * A function that allows to programmatically create the Astro i18n middleware.
+ *
+ * This is use useful when you still want to use the default i18n logic, but add only few exceptions to your website.
+ *
+ * ## Examples
+ *
+ * ```js
+ * // middleware.js
+ * import { middleware } from "astro:i18n";
+ * import { sequence, defineMiddleware } from "astro:middleware";
+ *
+ * const customLogic = defineMiddleware(async (context, next) => {
+ *   const response = await next();
+ *
+ *   // Custom logic after resolving the response.
+ *   // It's possible to catch the response coming from Astro i18n middleware.
+ *
+ *   return response;
+ * });
+ *
+ * export const onRequest = sequence(customLogic, middleware({
+ * 	prefixDefaultLocale: true,
+ * 	redirectToDefaultLocale: false
+ * }))
+ *
+ * ```
+ */
+export const middleware =
+	i18n?.routing === 'manual'
+		? (customOptions?: Options) => {
```

Comment:
I find this ternary and nesting to be a little difficult to read. Would it make sense instead to have this function be defined as `manualMiddleware` or something and then make the conditional a regular `if` statement?